### PR TITLE
Add the support for admins to specify designated matchmaking times.

### PR DESCRIPTION
**Pull request by 2Pacalypse-**
This PR allows admins to create new matchmaking times (and delete the
ones in future). The functionality to actually use this information and
notify the user when the matchmaking state changes will come in a next PR.

### CODE REVIEW

**Review by tec27 — CHANGES_REQUESTED** [review]

File: server/lib/models/matchmaking-times.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    enabled: props.enabled,
+  }
+}
+
+/**
+ * Retrieves a history of matchmaking times for a particular matchmaking type. History size can be
+ * provided which will be used to retrieve (at most) that amount of history on both sides of the
+ * date nearest to the current.
+ */
+export async function getMatchmakingTimesHistory(
```
[the remark this answers is not in the document]
Comment (2Pacalypse-) [reply server/lib/models/matchmaking-times.ts:33]: Actually, this query fetches the current applicable row, 10 future rows _and_ 10 past rows. That's mainly the reason for the complexity behind this query. I was going by the [spec](https://github.com/ShieldBattery/ShieldBattery/issues/336), which says:
- Admin UI could display next N future changes + M previous changes

Granted, N = M in this implementation :p
But yeah, I was thinking about this quite a bit (should've probably asked for clarification though), and thought it was fine to not have pagination in this case. Especially since this is only used to display a sort of snapshot of matchmaking times for admins when they're adding new times. In my next PR, which implements the user-facing things, I actually wrote a new query which fetches the future N rows, but also filters the ones that have the same `enabled` status as the one before them.

Anyway, I can split this query into multiple ones if that's what we want to do, but implementing a pagination for a list which shows current, future _and_ past rows might be a little bit tricky ^^. Unless we just stop trying to center around the current row and just get all the rows, in order and paginated. This is what we do in matchmaking map pools history btw. But there it's less likely we'll have a lot of entries in the future, unlike here, where matchmaking times might be entered by an admin well into the future.

File: server/lib/models/matchmaking-times.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  id: string
+  matchmaking_type: MatchmakingType
+  start_date: number
+  enabled: boolean
+  /* eslint-enable camelcase */
+}): MatchmakingTime {
+  return {
+    id: props.id,
+    type: props.matchmaking_type,
+    startDate: +props.start_date,
```
[the remark this answers is not in the document]
Comment (2Pacalypse-) [reply server/lib/models/matchmaking-times.ts:23]: Hmm, I'm not sure; it's definitely not a number so type on `props` was incorrect. It looks like a `Date`, but I don't know in which formatting.

I think returning it as a `number` makes more sense, since users usually create a new date out of it to format it to their liking (I'm not sure if this `Date` that's returned by postgres could be directly formatted). Also, I feel like comparing different dates (which is done in multiple places) is easier in their number form instead of their formatted `Date` form, but I could be wrong about this?

File: server/lib/models/matchmaking-times.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tec27) [reply server/lib/models/matchmaking-times.ts:33]: When I say stuff like that, I don't intend it to mean "this must be implemented as a single query on the backend." The fact that the UI is displaying that stuff just means that it must be query-able, not that it needs to be query-able as a single unit (and making it a single unit makes it really hard to do things like pagination, as you have discovered). I don't think displaying *only* N future and M previous changes is really tenable, there are going to be times where an admin needs to look up e.g. some far-past matchmaking time in order to check when something happened, or we have a lot of future times we need to look through, etc.

I would instead separate all these queries out: 
1) get the current time
2) get N times before time X
3) get N times after time X

The HTTP API function could call those 3 functions in that way (and parallelize 2 and 3), and 2 and 3 would be available for separate API calls to handle pagination stuff.

File: server/lib/models/matchmaking-times.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (tec27) [reply server/lib/models/matchmaking-times.ts:23]: sorry, I should have been more clear in my language. By "users of this API", I mean "callers of this function", not "users of the HTTP API that eventually calls into this function". `pg` returns `Date` objects for date columns, and those objects have a time zone associated with them, so by converting them to a number, you're throwing that information away. I understand that you eventually want to show these as a String on the client, but generally you want to keep data in it's most universal form until you need it in some other, specific form. e.g. if we wanted to display a number rounded to 2 decimal places on the client, we wouldn't do that rounding until the display layer, our database layer would return the full, unrounded form.

When transmitting `Date`s to clients through the HTTP API, the objects get coerced into a `number` anyway, as JSON does not have a representation for `Date`s. But if we need to do operations on these dates on the server before doing that transmission, or if we want to do things unrelated to transmitting them (e.g. set a timeout for the time between the current date and the one stored in the database, taking time zones into account), having the actual `Date` object is important.

So in short, I definitely think the thing in `MatchmakingTime` should be a `Date`, and you should not be doing this type coercion here.

(and as a side note, since some of your comment here leads me to believe you might be confused on this: `Date` is not a formatted date string, it's an object based around a timestamp + time zone. It can be compared to other dates properly (a later date is `>` an earlier date, for instance), you can subtract 2 dates and get back a delta in milliseconds, etc.)

File: server/lib/models/matchmaking-times.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (2Pacalypse-) [reply server/lib/models/matchmaking-times.ts:33]: Yeah, that sounds good. I just have a quick question. How should we display the history now in the admin panel? Just display the current time separately (somewhere on top) and then two tables for past/future times, each with their own paging?

File: server/lib/models/matchmaking-times.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (2Pacalypse-) [reply server/lib/models/matchmaking-times.ts:23]: Oh yeah, I misunderstood that. I was thinking of what we send over HTTP back to the clients. But even in that case, I wasn't aware of `Date`'s incompatibility with JSON ^^.

But yeah, all of that makes sense so I'll do it like that. Btw, the only reason why I even did this type coercion here is because we did it before on all of the things that are saved as `Date` in the database, afaik. I never really thought about it much, but might be useful to rethink those use cases as well.

[the remark this answers is not in the document]
Comment (tec27) [reply server/lib/models/matchmaking-times.ts:23]: I'm not sure what exactly you're looking at, because on a quick look I can only find things that are using Dates as Dates. (`users`, `games`, `chat-channels`, for example, all use Dates the way I described). Perhaps you wrote some code like this a while back and we didn't catch it then?

File: server/lib/models/matchmaking-times.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tec27) [reply server/lib/models/matchmaking-times.ts:33]: I'd probably make a table that starts with some rows on each side of the current time, maybe highlights the current time in some way, and has buttons or something to retrieve more rows on either side.

File: server/lib/models/matchmaking-times.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (2Pacalypse-) [reply server/lib/models/matchmaking-times.ts:23]: Nvm, you're right. I was thinking of stuff we do in HTTP/WS APIs, like here:
https://github.com/ShieldBattery/ShieldBattery/blob/master/server/lib/wsapi/chat.js#L149
https://github.com/ShieldBattery/ShieldBattery/blob/master/server/lib/api/bans.js#L49

According to the above, this is not really necessary either since those coercions will be done automatically because of JSON's incompatibility with the `Date` object?

[the remark this answers is not in the document]
Comment (tec27) [reply server/lib/models/matchmaking-times.ts:23]: I checked just to be sure and I was actually slightly wrong on the conversion. They *do* get converted when JSON.stringify'd, but to a standardized string in UTC (e.g. `'2020-12-17T01:05:53.951Z'`). This is probably slightly better than the number/timestamp form, since you can still use this to initialize a `Date` object and the format is self-documenting about the time zone, but probably isn't safe to change code that is currently doing the number conversion directly without looking at what's using it.

[the remark this answers is not in the document]
Comment (2Pacalypse-) [reply server/lib/models/matchmaking-times.ts:23]: After actually using it now, it's pretty annoying that you can't use this string directly in the time-related APIs after receiving it on the client, and you pretty much _have_ to convert it to a `Date` object first (preferably immediately in the reducer). That was a nice benefit of using the numbers format imo, since they can be transferred through JSON without any type conversion.

File: client/admin/matchmaking-times.jsx [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+          <p>
+            Something went wrong while trying to retrieve the matchmaking times history. The error
+            message was:
+          </p>
+          <ErrorText as='p'>{history.lastError.message}</ErrorText>
+        </>
+      )
+    }
+
+    if (history.sortedList.isEmpty()) {
```
Comment (tec27) [inline client/admin/matchmaking-times.jsx:134]: For a demonstrations as to why you shouldn't put expensive stuff in getters, look at how many times you use `history.sortedList` in this function. It doesn't *feel* expensive, it just looks like a variable access. But every access here is creating an entirely new list, sorting it, then reversing it. And that happens even if you're just checking if there's at least one entry. If it had been a function you probably would have used it once and stored the value, then reused that value throughout the function.

File: server/lib/api/matchmakingTimes.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  addMatchmakingTime,
+  getMatchmakingTimeById,
+  removeMatchmakingTime,
+} from '../models/matchmaking-times'
+
+const matchmakingTypeSchema = Joi.object({
+  matchmakingType: Joi.valid(...MATCHMAKING_TYPES).required(),
+})
+
+const addMatchmakingTimeSchema = Joi.object({
```
Comment (tec27) [inline server/lib/api/matchmakingTimes.ts:26]: I think you might need to make this entire object required. You can check by trying a request that does not pass a request body

File: server/lib/api/matchmakingTimes.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  startDate: Joi.number().integer().greater(Date.now()).required(),
```
Comment (tec27) [inline server/lib/api/matchmakingTimes.ts:27]: I don't think this validation doesn't what you want it to. `Date.now` will be evaluated once, when this file is evaluated, and never again for the duration of the process. You probably want to use the dynamic validation stuff Joi has instead

File: server/lib/models/matchmaking-times.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    FROM matchmaking_times
+    WHERE matchmaking_type = ${matchmakingType} AND start_date
+  `
+    .append(isFuture ? '>' : '<')
+    .append(sql`${date};`)
+
+  const { client, done } = await db()
+  try {
+    const result = await client.query(query)
+    return parseInt(result.rows[0].count, 10)
```
Comment (tec27) [inline server/lib/models/matchmaking-times.ts:64]: I don't think this is a string?

(You seem to have done this and the `COUNT(id)` thing in a few places in other models, we should fix those as well)

File: server/lib/models/matchmaking-times.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  }
+}
+
+async function getTotalTimes(
+  matchmakingType: MatchmakingType,
+  date = new Date(),
+  isFuture = true,
+): Promise<number> {
+  const query = sql`
+    SELECT COUNT(id)
```
Comment (tec27) [inline server/lib/models/matchmaking-times.ts:54]: Generally I think you're better off just doing `COUNT(*)`. I assume you're doing `COUNT(id)` under an assumption that it will be faster, since it names less columns, but `COUNT(*)` is actually parsed as a specific thing and doesn't require the DB to look at every row, whereas your query *does* require looking at the rows.

File: server/lib/models/matchmaking-times.ts [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    return result.rows.length > 0 ? convertFromDb(result.rows[0]) : null
+  } finally {
+    done()
+  }
+}
+
+async function getTotalTimes(
+  matchmakingType: MatchmakingType,
+  date = new Date(),
+  isFuture = true,
```
Comment (tec27) [inline server/lib/models/matchmaking-times.ts:51]: Use an enum for this instead, the fact that you need to comment every time you set this parameter what it actually means should be a good indication that it's not a particularly clear or safe way of specifying this

File: server/migrations/20201212153557-add-index-to-matchmaking-times.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+exports.up = async function (db) {
+  await db.runSql(`
+    CREATE INDEX start_date_index ON matchmaking_times (start_date);
```
Comment (tec27) [inline server/migrations/20201212153557-add-index-to-matchmaking-times.js:3]: I'm not sure this will actually be used for the query we will be running most often (the current matchmaking time). I think you'd need to store it in the reverse order for that (which is what we do for chat messages), e.g.

```sql
CREATE INDEX start_date_index ON matchmaking_times (start_date DESC);
```

**Review by tec27 — CHANGES_REQUESTED** [review]

File: server/lib/api/matchmakingTimes.ts (hunk 4 above)
Comment (2Pacalypse-) [reply server/lib/api/matchmakingTimes.ts:26]: Hmm, it does seem to work without that from what I can tell. Passing no body, I get this on the client:

image

File: server/lib/models/matchmaking-times.ts (hunk 6 above)
Comment (2Pacalypse-) [reply server/lib/models/matchmaking-times.ts:64]: Hmm, `typeof` does detect it as a `string` from what I can tell.

File: server/lib/api/matchmakingTimes.ts (hunk 4 above)
Comment (tec27) [reply server/lib/api/matchmakingTimes.ts:26]: Cool, just wanted to make sure since I saw some weirdness but it was with requests made from Rust, and I chose to err on the safer side rather than recompiling/restarting games a bunch of times to find out :)

File: server/lib/models/matchmaking-times.ts (hunk 6 above)
Comment (tec27) [reply server/lib/models/matchmaking-times.ts:69]: Yeah I guess that is actually the case:

> Unfortunately those are going to return strings now since the return type of the COUNT operator is bigint which, by definition, can exceed the maximum value of an int in JavaScript.

We probably need to fix the other places that use this now, where it seems to work by accident. I guess it looks like just email verifications now that I look at it. `invites` returns a badly typed thing from one of its queries but fixes it at some other random point, but that's soon-to-be-deleted code anyway.

**Review by tec27 — APPROVED** [review]
